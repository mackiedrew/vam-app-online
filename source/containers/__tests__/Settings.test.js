import {
  Settings as Subject,
  mapStateToProps,
  mapDispatchToProps
} from "../Settings";

const mockSettings = {
  grain: {
    label: "Grain Time",
    value: 5,
    unit: "s",
    type: "number"
  },
  quietCutoff: {
    label: "Quietness Threshold",
    value: 10,
    unit: "%",
    type: "number"
  }
};

const mockHotkeys = {
  play: {
    label: "Play/Pause Key",
    value: "p"
  },
  next: {
    label: "Next Grain Key",
    value: "e"
  },
  previous: {
    label: "Previous Grain Key",
    value: "q"
  },
  nextTrack: {
    label: "Next Track Key",
    value: "j"
  },
  previousTrack: {
    label: "Previous Track Key",
    value: "k"
  }
};

const mockProps = {
  open: true,
  settings: mockSettings,
  hotkeys: mockHotkeys
};

describe("<Settings />", () => {
  it("renders without crashing", () => {
    const subject = shallow(<Subject {...mockProps} />);
    expect(subject.is("aside.settings")).toBe(true);
  });

  it("renders with open class when passed open prop as true", () => {
    const subject = shallow(<Subject {...mockProps} open />);
    expect(subject.is("aside.open")).toBe(true);
  });

  it("renders with closed class when passed open prop as false", () => {
    const subject = shallow(<Subject {...mockProps} open={false} />);
    expect(subject.is("aside.closed")).toBe(true);
  });

  describe("renders correctly", () => {
    it("while open", () => {
      const tree = renderer.create(<Subject {...mockProps} open />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("while closed", () => {
      const tree = renderer
        .create(<Subject {...mockProps} open={false} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it("triggers changeHotkey when handleHotKeyChange is called", () => {
    const mockChangeHotkey = sinon.spy();
    const mockPreventDefault = sinon.spy();
    const mockEvent = {
      preventDefault: mockPreventDefault,
      target: {
        value: 50
      }
    };
    const subject = shallow(
      <Subject {...mockProps} changeHotkey={mockChangeHotkey} />
    );
    subject.instance().handleHotkeyChange(mockEvent, "testOperation");

    expect(mockPreventDefault.called).toBe(true);
    expect(mockChangeHotkey.called).toBe(true);
  });

  it("triggers changeSetting when handleSettingsChange is called", () => {
    const mockChangeSetting = sinon.spy();
    const mockPreventDefault = sinon.spy();
    const mockEvent = {
      preventDefault: mockPreventDefault,
      target: {
        value: 50
      }
    };
    const subject = shallow(
      <Subject {...mockProps} changeSetting={mockChangeSetting} />
    );
    subject.instance().handleSettingsChange(mockEvent, "testSetting");

    expect(mockPreventDefault.called).toBe(true);
    expect(mockChangeSetting.called).toBe(true);
  });

  it("triggers pauseControls when handleFocus is called", () => {
    const mockPauseControls = sinon.spy();
    const subject = shallow(
      <Subject {...mockProps} pauseControls={mockPauseControls} />
    );
    subject.instance().handleFocus();
    expect(mockPauseControls.called).toBe(true);
  });

  it("triggers resumeControls when handleBlur is called", () => {
    const mockResumeControls = sinon.spy();
    const subject = shallow(
      <Subject {...mockProps} resumeControls={mockResumeControls} />
    );
    subject.instance().handleBlur();
    expect(mockResumeControls.called).toBe(true);
  });

  describe("mapStateToProps()", () => {
    const mockState = {
      ui: {
        settingsOpen: 1
      },
      settings: 2,
      keyboard: {
        hotkeys: 3
      }
    };

    it("returns object with expected properties", () => {
      const { open, settings, hotkeys } = mapStateToProps(mockState);
      expect(open).toBe(1);
      expect(settings).toBe(2);
      expect(hotkeys).toBe(3);
    });
  });

  describe("mapDispatchToProps()", () => {
    it("returns object with expected properties", () => {
      const mockDispatch = sinon.spy();
      const {
        changeSetting,
        changeHotkey,
        resumeControls,
        pauseControls
      } = mapDispatchToProps(mockDispatch);
      expect(typeof changeSetting).toBe("function");
      expect(typeof changeHotkey).toBe("function");
      expect(typeof resumeControls).toBe("function");
      expect(typeof pauseControls).toBe("function");
    });
  });
});
