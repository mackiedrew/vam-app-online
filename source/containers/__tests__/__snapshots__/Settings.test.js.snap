// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> renders correctly while closed 1`] = `
<aside
  className="settings closed"
>
  <fieldset>
    <div
      className="settings-field"
    >
      <label
        htmlFor="grain"
      >
        Grain Time
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="grain"
          name="grain"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="number"
          value={5}
        />
        <span
          className="unit"
        >
          s
        </span>
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="quietCutoff"
      >
        Quietness Threshold
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="quietCutoff"
          name="quietCutoff"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="number"
          value={10}
        />
        <span
          className="unit"
        >
          %
        </span>
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="play"
      >
        Play/Pause Key
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="play"
          name="play"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type={undefined}
          value="p"
        />
        <span
          className="unit"
        />
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="next"
      >
        Next Grain Key
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="next"
          name="next"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type={undefined}
          value="e"
        />
        <span
          className="unit"
        />
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="previous"
      >
        Previous Grain Key
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="previous"
          name="previous"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type={undefined}
          value="q"
        />
        <span
          className="unit"
        />
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="nextTrack"
      >
        Next Track Key
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="nextTrack"
          name="nextTrack"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type={undefined}
          value="j"
        />
        <span
          className="unit"
        />
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="previousTrack"
      >
        Previous Track Key
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="previousTrack"
          name="previousTrack"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type={undefined}
          value="k"
        />
        <span
          className="unit"
        />
      </div>
    </div>
  </fieldset>
</aside>
`;

exports[`<Settings /> renders correctly while open 1`] = `
<aside
  className="settings open"
>
  <fieldset>
    <div
      className="settings-field"
    >
      <label
        htmlFor="grain"
      >
        Grain Time
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="grain"
          name="grain"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="number"
          value={5}
        />
        <span
          className="unit"
        >
          s
        </span>
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="quietCutoff"
      >
        Quietness Threshold
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="quietCutoff"
          name="quietCutoff"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="number"
          value={10}
        />
        <span
          className="unit"
        >
          %
        </span>
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="play"
      >
        Play/Pause Key
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="play"
          name="play"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type={undefined}
          value="p"
        />
        <span
          className="unit"
        />
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="next"
      >
        Next Grain Key
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="next"
          name="next"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type={undefined}
          value="e"
        />
        <span
          className="unit"
        />
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="previous"
      >
        Previous Grain Key
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="previous"
          name="previous"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type={undefined}
          value="q"
        />
        <span
          className="unit"
        />
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="nextTrack"
      >
        Next Track Key
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="nextTrack"
          name="nextTrack"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type={undefined}
          value="j"
        />
        <span
          className="unit"
        />
      </div>
    </div>
    <div
      className="settings-field"
    >
      <label
        htmlFor="previousTrack"
      >
        Previous Track Key
      </label>
      <div
        className="entry-area"
      >
        <input
          className="input"
          id="previousTrack"
          name="previousTrack"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type={undefined}
          value="k"
        />
        <span
          className="unit"
        />
      </div>
    </div>
  </fieldset>
</aside>
`;
